### PR TITLE
Added the ability to pass arbitrary headers.

### LINKED ISSUE #186: Allow passing arbitrary headers
Support for a -H option (like curl) to support authentication, proxies, etc.

### PR DESCRIPTION
**Pull request by omadawn**
Added a --header | -H option to allow the passing of arbitrary headers.

Fixes #186

### CODE REVIEW

File: modules/install-consul/install-consul [hunk 1]
```diff
@@ -164,14 +168,26 @@ function create_consul_install_paths {
...
   if [[ -z "$download_url" && -n "$version" ]];  then
     download_url="https://releases.hashicorp.com/consul/${version}/consul_${version}_linux_amd64.zip"
   fi
 
+  if [[ ! -z "$addtl_headers" ]]; then
+    local headers
+
+    for value in "${addtl_headers[@]}"
+    do
+      headers+="-H \"$value\" "
```
Comment (brikis98) [inline modules/install-consul/install-consul:185]: Slightly cleaner to make `headers` an array (declare it up above `local -a headers=()`), appending to it here via `headers+=("-H \"$value\"")`, and expanding it in the `curl` command below with `${headers[@]}`.

**Review by brikis98 — COMMENTED** [review]
Thanks for the PR!

File: modules/install-consul/install-consul [hunk 2]
```diff
@@ -25,13 +27,15 @@ function print_usage {
...
   echo "Options:"
   echo
-  echo -e "  --version\t\tThe version of Consul to install. Optional if download-url is provided."
-  echo -e "  --download-url\t\tUrl to exact Consul package to be installed. Optional if version is provided."
-  echo -e "  --path\t\tThe path where Consul should be installed. Optional. Default: $DEFAULT_INSTALL_PATH."
-  echo -e "  --user\t\tThe user who will own the Consul install directories. Optional. Default: $DEFAULT_CONSUL_USER."
-  echo -e "  --ca-file-path\t\tPath to a PEM-encoded certificate authority used to encrypt and verify authenticity of client and server connections. Will be installed under <install-path>/tls/ca."
-  echo -e "  --cert-file-path\t\tPath to a PEM-encoded certificate, which will be provided to clients or servers to verify the agent's authenticity. Will be installed under <install-path>/tls. Must be provided along with --key-file-path."
-  echo -e "  --key-file-path\t\tPath to a PEM-encoded private key, used with the certificate to verify the agent's authenticity. Will be installed under <install-path>/tls. Must be provided along with --cert-file-path"
+  echo -e "  --help | -h\t\tPrints this usage message."
```
[the remark this answers is not in the document]
Comment (omadawn) [reply modules/install-consul/install-consul:30]: I'll try and get to that this weekend.

File: modules/install-consul/README.md [hunk 3]
```diff
@@ -44,20 +44,25 @@ example](https://github.com/hashicorp/terraform-aws-consul/tree/master/examples/
...
 
-Example:
+Examples:
 
 ```
 install-consul --version 1.2.2
 ```
 
+```
+install-consul --download-url https://staticassets.example.com/consul/1.2.3/consul_1.2.3_linux_amd64.zip -H "X-API-Key: 123fks9213432"
```
[the remark this answers is not in the document]
Comment (omadawn) [reply modules/install-consul/README.md:64]: We have no idea what customers would use for storing static assets. My use case actually requires the API key in the header `X-JFrog-Art-Api` but that would be different for nexus or some generic FTP server.
On top of that a generic header is more flexible so we wouldn't need another PR when someone wants to add a header for anything else.

If the header was standard across platforms we could add both but I didn't want to add any specifics for a single unrelated technology.

[the remark this answers is not in the document]
Comment (brikis98) [reply modules/install-consul/README.md:64]: Got it. I was wondering if HashiCorp/Consul itself had some native header it was requiring now (e.g., for Consul Enterprise), but if that's not the case, sticking with a generic header makes sense.

**Pull request closed without merging**